modules/standard/DistributedDeque.chpl

```diff
@@ -0,0 +1,1000 @@
+      return;
+    }
+
+    // We iterate directly over the heads of each slot, so we capture them in advance.
+    var nodes : [{0..#nSlots}] (int, int, LocalDequeNode(eltType));
+    for i in 0 .. #nSlots {
+      var node = slots[i].head;
+      nodes[i] = (node.size, node.headIdx, node);
+    }
+
+    // Iterate over captured head nodes; each time we read them we advance them
+    var iterations = size;
+    for i in head..#tail {
+      if iterations == 0 then break;
+      iterations -= 1;
+
+      var idx = i % nSlots;
+      var (size, headIdx, node) = nodes[idx];
+      if node == nil {
+        halt("Iterating over nil nodes, head: ", head, ", tail: ", tail, ", idx: ", i);
+      }
+      yield node.elements[headIdx];
+
+      // Update state...
+      size -= 1;
+      headIdx += 1;
+      if headIdx > DEQUE_BLOCK_SIZE {
+        headIdx = 1;
+      }
+
+      // Advance...
+      if size == 0 {
+        node = node.next;
+        if node != nil {
+          nodes[idx] = (node.size, node.headIdx, node);
+        } else {
+          nodes[idx] = (-1, -1, node);
+        }
+      } else {
+        // Else update state...
+        nodes[idx] = (size, headIdx, node);
+      }
+    }
+  }
+
+  /*
+    Iterates over the deque in Last-In-First-Out order, from back to front. The
+    deque must be frozen or it will result in a halt. This operation is sequential.
+  */
+  iter LIFO() : eltType {
+    if !isFrozen() {
+      halt("Ordered iteration requires the queue to be frozen.");
+    }
+
+    var localThis = getPrivatizedThis;
+
+    // Fill our slots to visit in FIFO order.
+    var head = globalHead.read();
+    var tail = globalTail.read();
+    var size = queueSize.read();
+
+    // Check if empty...
+    if size == 0 {
+      return;
+    }
+
+    // We iterate directly over the heads of each slot, so we capture them in advance.
+    var nodes : [{0..#nSlots}] (int, int, LocalDequeNode(eltType));
+    for i in 0 .. #nSlots {
+      var node = slots[i].tail;
+      nodes[i] = (node.size, node.tailIdx, node);
+    }
+
+    // Iterate over captured head nodes; each time we read them we advance them
+    var iterations = size;
+    while tail - 1 >= head - 1 {
+      if iterations == 0 then break;
+      iterations -= 1;
+
+      var idx = (tail - 1) % nSlots;
+      var (size, tailIdx, node) = nodes[idx];
+      if node == nil {
+        halt("Iterating over nil nodes, head: ", head, ", tail: ", tail, ", idx: ", tail-1);
+      }
+      yield node.elements[tailIdx];
+
+      // Update state...
+      size -= 1;
+      tailIdx -= 1;
+      if tailIdx == 0 {
+        tailIdx = DEQUE_BLOCK_SIZE;
+      }
+
+      // Advance...
+      if size == 0 {
+        node = node.prev;
+        if node != nil {
+          nodes[idx] = (node.size, node.tailIdx, node);
+        } else {
+          nodes[idx] = (-1, -1, node);
+        }
+      } else {
+        // Else update state...
+        nodes[idx] = (size, tailIdx, node);
+      }
+    }
+  }

```

Comment:
Yeah, I was thinking of having `these(param order : orderKind = orderKind.NONE)`, but I wasn't sure. That way I'd have `orderKind.NONE, orderKind.FIFO, orderKind.LIFO`, etc. As well, since enums no longer causes a compiler internal error with deadCodeElimination (fixed in #6570), I have no issue with enums anymore.